### ISSUE TITLE
Price LP NFT's (research needed)

### ISSUE THREAD

**Issue by grod220** [penumbra-zone/web#528]
For LP NFT's, populate ValueView's equivalentValues array with pricing data:

- Find a way to convert LP position into underlying tokens (research needed)
- Add to equivalentValues array

**Comment by jessepinho** [comment]
Can you elaborate a bit?

1. What query would we make to the fullnode? i.e., what service would we hit?
2. What does "the current reserves" mean, and what does "a pair of assets" refer to?

**Comment by grod220** [comment]
> What query would we make to the fullnode? i.e., what service would we hit?

We'd query `https://grpc.testnet.penumbra.zone/` with likely this message: https://buf.build/penumbra-zone/penumbra/docs/main:penumbra.core.component.dex.v1#penumbra.core.component.dex.v1.QueryService.LiquidityPositionById

> What does "the current reserves" mean, and what does "a pair of assets" refer to?

For a DEX, users can come and make trades (X asset for Y). But to do so, there must be other participants to provide liquidity for that trade. They are called "liquidity providers" or LPs. In exchange for this, they receive trading fees. So pools (say OSMO-UM) represent the aggregated liquidity for trading between those two assets. So when someone has LP'd they'll get LP tokens back. To price this, we must query to see what the underlying assets it represents (e.g. 123 LP tokens = 10 OSMO and 9932 UM). We can then convert both of those to the numeraire (like USDC) and show that to the user to represent the LP position's current value.

**Comment by jessepinho** [comment]
Per today's web weekly, moving this back to backlog as it's low-prio

**Comment by hdevalence** [comment]
Yes, the LP side of the DEX basically assumes that a user is talking to a trusted RPC.  If a user has LP NFTs, there's no way to get detailed info without talking to an RPC server.

But, we're expecting that most users of minifront will be on the swapper side, rather than the MM side, so this is pretty low priority.

**Comment by jessepinho** [comment]
@grod220, following up about this. You wrote:

> To price this, we must query to see what the underlying assets it represents (e.g. 123 LP tokens = 10 OSMO and 9932 UM). We can then convert both of those to the numeraire (like USDC) and show that to the user to represent the LP position's current value.

A few questions:

1. It'd just be 1 LP token for a given liquidity position, right? You wrote 123 LP tokens in your example, but actually, a person only every owns 1 LP token for a given LP, right? For example, I just ran the command ` pcli tx position order buy 1penumbra@5test_usd` twice in a row in my console. Instead of getting two of the same LpNFTs, I got one each of two different LpNFTs. Just wanted to confirm I'm understanding this correctly?
2. Let's say I've opened a position to buy 1000 UM at 5 OSMO per UM. I have spent 5000 OSMO, and I get back 1 LP NFT.
  a. At the beginning, is my NFT worth 5000 OSMO and 0 UM?
  b. Let's say someone swaps 900 UM for OSMO. Now (if I understand correctly how liquidity works), there's 100 UM left to fill in my liquidity position. Is my NFT now worth 900 UM and 500 OSMO?
3. Do we want to show the user the value of their NFT A) in terms of the underlying assets, B) in terms of the _numeraires_ of their underlying assets, or C) both? 

    For example, if we go with option C, it'd look something like this (assuming a 1:1 ratio between UM <-> USD and OSMO <-> USD, for simplicity):

    ```
    Assets        | Equivalent values
    -----------------------------------------------------
    1 LP NFT      | 900 UM ($900USD) / 500 OSMO ($500USD)
    ```

**Comment by jessepinho** [comment]
CC @erwanor for questions 1 and 2, since he knows a lot about dex.

**Comment by grod220** [comment]
> It'd just be 1 LP token for a given liquidity position, right? 

My previous example with multiple fungible LP tokens fits more with the _old_ way of providing liquidity (not relevant to Penumbra sorry). You may have heard of this being called v2 versus v3 or concentrated liquidity (referencing uniswap). With concentrated liquidity, everyone creates a unique LP position so it makes sense for a single NFT to be issued.

> Let's say I've opened a position to buy 1000 UM at 5 OSMO per UM

I'm not sure that's quite the right wording. Play around with [Uniswap's v3 LP'ing tool](https://app.uniswap.org/add/ETH/0xB8c77482e45F1F44dE1745F52C74426C631bDD52/10000?minPrice=2.170290&maxPrice=8.681159). I've found [this explainer video](https://www.youtube.com/watch?v=Ehm-OYBmlPM) helpful too. You specify a range and amounts you'd like to provide to the pool to be swapped through. The exact math is a bit beyond me, but for our code, we don't need to do any math. We should just be able to query a node and its responsible for doing the math to simulate a withdraw and give us the underlying estimate.

> Do we want to show the user the value of their NFT...

I'd say A or C makes most sense to me.
